File: src/session.ts
import { FirefoxProcessManager } from './process-manager.js';
import { BiDiConnection } from './bidi-connection.js';
import { BiDiPage } from './bidi-page.js';
import { setLogger, ConsoleLogger } from './logger.js';
import { ChromeLauncher } from './chrome/index.js';
import type { FirefoxLaunchOptions, ChromeLaunchOptions } from './types.js';

export type SessionOptions = FirefoxLaunchOptions | ChromeLaunchOptions;

export class Session {
  readonly page: BiDiPage;

  constructor(
    readonly connection: BiDiConnection,
    private readonly manager: FirefoxProcessManager | null,
    initialContextId: string,
    private readonly onClose?: () => Promise<void>
  ) {
    this.page = new BiDiPage(connection, initialContextId);
  }

  /** Wrap an existing browsing context — use for iframes, child contexts, etc. */
  getPage(contextId: string): BiDiPage {
    return new BiDiPage(this.connection, contextId);
  }

  /** Open a new top-level browsing context (new tab). */
  async newPage(): Promise<BiDiPage> {
    const result = await this.connection.sendCommand('browsingContext.create', { type: 'tab' });
    return new BiDiPage(this.connection, result.context);
  }

  async close(): Promise<void> {
    await this.connection.close();
    await this.manager?.kill();
    await this.onClose?.();
  }
}

export async function startSession(options: SessionOptions = {}): Promise<Session> {
  if (process.env.BIDI_DEBUG === '1') {
    setLogger(new ConsoleLogger(true));
  }

  const browser = options.browser ?? (process.env.BIDI_BROWSER as 'firefox' | 'chrome' | undefined) ?? 'firefox';

  if (browser === 'chrome') {
    const opts = options as ChromeLaunchOptions;

    const headless = opts.headless ?? (process.env.CHROME_HEADLESS === '1');
    const biDiUrlOption = opts.biDiUrl ?? process.env.CHROME_BIDI_URL;
    const chromePathOption = opts.chromePath ?? process.env.CHROME_PATH;

    if (biDiUrlOption) {
      // Raw BiDi URL — user manages everything
      const connection = new BiDiConnection();
      await connection.connect(biDiUrlOption);
      try {
        await connection.sendCommand('session.new', { capabilities: {} });
      } catch { /* session already exists */ }
      const tree = await connection.sendCommand('browsingContext.getTree', {});
      const initialContextId = tree.contexts[0].context;
      return new Session(connection, null, initialContextId);
    }

    // Auto-launch Chrome with BidiServer bridged directly (no WebSocket indirection)
    const launcher = new ChromeLauncher();
    const connection = new BiDiConnection();
    await launcher.launch(connection, {
      chromePath: chromePathOption,
      headless,
      args: opts.args,
    });

    await connection.sendCommand('session.new', { capabilities: {} });
    const tree = await connection.sendCommand('browsingContext.getTree', {});
    const initialContextId = tree.contexts[0].context;

    return new Session(connection, null, initialContextId, () => launcher.kill());
  }

  // Firefox (default)
  const opts = options as FirefoxLaunchOptions;
  const resolvedOptions: FirefoxLaunchOptions = {
    ...opts,
    firefoxPath: opts.firefoxPath ?? process.env.FIREFOX_PATH,
    profilePath: opts.profilePath ?? process.env.FIREFOX_PROFILE,
    headless:    opts.headless    ?? (process.env.FIREFOX_HEADLESS === '1'),
    prefs:       opts.prefs       ?? (process.env.FIREFOX_PREFS ? JSON.parse(process.env.FIREFOX_PREFS) : undefined),
  };

  const manager = new FirefoxProcessManager();
  const connection = new BiDiConnection();

  const port = await manager.launch(resolvedOptions);
  await connection.connect(port);

  const alwaysMatch: Record<string, any> = {};
  if (resolvedOptions.acceptInsecureCerts) {
    alwaysMatch.acceptInsecureCerts = true;
  }
  if (resolvedOptions.prefs && Object.keys(resolvedOptions.prefs).length > 0) {
    alwaysMatch['moz:firefoxOptions'] = { prefs: resolvedOptions.prefs };
  }
  await connection.sendCommand('session.new', { capabilities: { alwaysMatch } });

  const tree = await connection.sendCommand('browsingContext.getTree', {});
  const initialContextId = tree.contexts[0].context;

  return new Session(connection, manager, initialContextId);
}

File: python/firefox_bidi_client/session.py
import os
from typing import Optional

from .bidi_connection import BiDiConnection
from .bidi_page import BiDiPage
from .logger import ConsoleLogger, set_logger
from .process_manager import FirefoxProcessManager


class Session:
    def __init__(
        self,
        connection: BiDiConnection,
        manager: FirefoxProcessManager,
        initial_context_id: str,
    ):
        self.connection = connection
        self._manager = manager
        self.page = BiDiPage(connection, initial_context_id)

    def get_page(self, context_id: str) -> BiDiPage:
        """Wrap an existing browsing context — use for iframes, child contexts, etc."""
        return BiDiPage(self.connection, context_id)

    async def new_page(self) -> BiDiPage:
        """Open a new top-level browsing context (new tab)."""
        result = await self.connection.send_command("browsingContext.create", {"type": "tab"})
        return BiDiPage(self.connection, result["context"])

    async def close(self) -> None:
        await self.connection.close()
        await self._manager.kill()

    async def __aenter__(self) -> "Session":
        return self

    async def __aexit__(self, *args) -> None:
        await self.close()


async def start_session(
    firefox_path: Optional[str] = None,
    headless: Optional[bool] = None,
    profile_path: Optional[str] = None,
    viewport: Optional[dict] = None,
    args: Optional[list] = None,
    accept_insecure_certs: bool = False,
    env: Optional[dict] = None,
    log_file: Optional[str] = None,
    prefs: Optional[dict] = None,
) -> Session:
    """Launch Firefox with BiDi and return a Session ready for automation.

    Environment variable overrides (when the corresponding kwarg is None):
        FIREFOX_PATH, FIREFOX_PROFILE, FIREFOX_HEADLESS=1, BIDI_DEBUG=1
    """
    if os.environ.get("BIDI_DEBUG") == "1":
        set_logger(ConsoleLogger(debug_enabled=True))

    options = {
        "firefox_path": firefox_path or os.environ.get("FIREFOX_PATH"),
        "headless": headless if headless is not None else (os.environ.get("FIREFOX_HEADLESS") == "1"),
        "profile_path": profile_path or os.environ.get("FIREFOX_PROFILE"),
        "viewport": viewport,
        "args": args,
        "accept_insecure_certs": accept_insecure_certs,
        "env": env,
        "log_file": log_file,
        "prefs": prefs,
    }

    manager = FirefoxProcessManager()
    connection = BiDiConnection()

    port = await manager.launch(options)
    await connection.connect(port)

    always_match: dict = {}
    if accept_insecure_certs:
        always_match["acceptInsecureCerts"] = True
    if prefs:
        always_match["moz:firefoxOptions"] = {"prefs": prefs}

    await connection.send_command("session.new", {"capabilities": {"alwaysMatch": always_match}})

    tree = await connection.send_command("browsingContext.getTree", {})
    initial_context_id = tree["contexts"][0]["context"]

    return Session(connection, manager, initial_context_id)
